OpenTerm/Controller/ViewController.swift

```diff
@@ -339,6 +357,261 @@ extension ViewController: UIDocumentPickerDelegate {
+                 *  Get all files that are in the bookmarks directory.
+                 */
+                let bookmarkFiles = try FileManager.default.contentsOfDirectory(atPath: bookmarkDirectoryURL.path)
+                
+                /**
+                 *  The array that will hold the obtained URLs. This will be returned eventually.
+                 */
+                var bookmarkURLs = [URL]()
+                
+                /**
+                 *  Iterate all bookmark filenames.
+                 */
+                for bookmarkFileName in bookmarkFiles {
+                    
+                    /**
+                     *  Get the url of the current bookmark file.
+                     */
+                    let bookmarkDataURL = bookmarkDirectoryURL.appendingPathComponent(bookmarkFileName)
+                    
+                    /**
+                     *  We try to load the bookmark from the file.
+                     */
+                    let loadedBookmark = try URL.bookmarkData(withContentsOf: bookmarkDataURL)
+                    
+                    /**
+                     *  This variable will indicate whether the bookmark is stale.
+                     *  If the bookmark is stale we create a new bookmark with the
+                     *  obtained URL and save the new one instead of the old one.
+                     */
+                    var isStale = true
+                    
+                    /**
+                     *  Try to obtain the URL from the bookmark.
+                     */
+                    if let loadedBookmarkURL = try URL(resolvingBookmarkData: loadedBookmark, bookmarkDataIsStale: &isStale) {
+                        
+                        /**
+                         *  If the bookmark is stale, we create a new bookmark
+                         *  from the obtained URL.
+                         */
+                        if isStale {
+                            do {
+                                try self.saveBookmarkURL(url: loadedBookmarkURL)
+                            }
+                        }
+                        
+                        /**
+                         *  Append the loaded URLs.
+                         */
+                        bookmarkURLs.append(loadedBookmarkURL)
+                    }
+                }
+                
+                return bookmarkURLs
+            } catch {
+                return []
+            }
+        }
+        
+        return []
+    }
+    
+    /// When the bookmark view controller did select a bookmark, we change the
+    /// current directory to the bookmarked url.
+    /// - Note: Only urls that were saved as bookmarks will work.
+    ///
+    /// - Parameter bookmarkURL: The bookmark that was selected.
+    func didSelectBookmarkURL(bookmarkURL: URL) {
+        
+        /**
+         *  Access the bookmarked URL
+         */
+        _ = bookmarkURL.startAccessingSecurityScopedResource()
```

Comment:
I believe that Apple recommends that this call needs to be balanced with a call to `stopAccessingSecurityScopedResource()`

Reply:
Yeah, I read that as well. Apparently there is also no balancing call to `startAccessingSecurityScopedResource` in `documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL])`. Any idea for a good place to place the `stopAccessingSecurityScopedResource()` calls? Does it makes sense to stopAccessing after changing the directory?

Reply:
I'm not positive how security scoping works with bookmarks & these start/stop methods. Apple says about stop: "When you call this method, you immediately lose access to the resource in question."

This doesn't invalidate the bookmark, does it?

If it doesn't invalidate the bookmark, I'd vote that we call start/stop when we enter/leave directories.
If it invalidates the bookmark, then call stop only on bookmark deletion.

Reply:
One way to approach this is to just add a `defer` call right after it, so that you'll know you have access to the resource in the same scope where you granted it - but not any longer. Means you might need to call it in more places, but at least it's clear and explicit that every function that needs to get to the resource has access to it.

So you could write here:

```
_ = bookmarkURL.startAccessingSecurityScopedResource()
defer { bookmarkURL.stopAccessingSecurityScopedResource() }
```

Reply:
I just tested and it looks like `stop` does not invalidate the bookmark. Is there a nice way to detect when we leave a directory?

Reply:
@nighthawk unfortunately, the scope ends only two statements later, so I don't think `defer` will work. 

For the bookmarks I could try accessing them when they are loaded from the file and stop accessing when the bookmark manager is de-initialized. But that won't help with the URL from the document picker.